torch/_functorch/partitioners.py

```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+    # Some example models to look at where this helps perf: poolformer_m36,
+    # mixer_b16_224, cait_m36_384
+
+    # The "rough" idea here is that if you have some node that is used by both a
+    # node nearby downstream as well as a node far downstream, if we recompute
+    # both of the downstream nodes, we're unlikely to be able to fuse both
+    # downstream nodes together.
+
+    # Thus, we shouldn't aim to recompute far downstream nodes that depend on
+    # this node. That intuition of "far downstream" is captured by whether
+    # there's an unfusible op along the chain somewhere
+
+    # It could probably be improved by properly analyzing what's going on in the
+    # backwards pass instead of only relying on whether it's unfusible in the
+    # forwards.
+
+    def find_first_unfusible(start_nodes: List[fx.Node], max_range: int) -> int:
+        """
+        Finds the first unfusible node in the chain of nodes starting from
+        `start_nodes` and returns its position.
+        """
+        sorted_nodes = []
+        for n in start_nodes:
+            heapq.heappush(sorted_nodes, (n.fw_order, n, True))
+
+        while len(sorted_nodes) > 0:
+            _, node, node_is_fusible = heapq.heappop(sorted_nodes)
+            if not node_is_fusible:
+                return node.fw_order
+            if node.fw_order > max_range:
+                continue
+            for user in node.users:
+                if user in required_fw_nodes:
+                    heapq.heappush(sorted_nodes, (user.fw_order, user, is_fusible(node, user)))
+        return max_range
+
+    if BAN_IF_USED_FAR_APART:
+        for used_node in required_fw_nodes:
+            orders = [user.fw_order for user in used_node.users if user in required_fw_nodes]
+            fw_users = [user for user in used_node.users if user in required_fw_nodes]
+            if len(orders) > 0:
+                first_unfusible_use = find_first_unfusible(fw_users, max(orders))
+                for user in tuple(used_node.users):
+                    if user in required_fw_nodes and user.fw_order > first_unfusible_use and is_fusible(used_node, user):
+                        if user in banned_nodes:
+                            continue
+                        log.info(
+                            "used above/below fusible %s:(%s) -> %s -> %s:(%s)",
+                            used_node, used_node.fw_order, first_unfusible_use, user, user.fw_order
+                        )
+                        ban_recomputation_if_allowed(user)
+
+
+    # This heuristic is fairly straightforward. The idea is that although it is
+    # cheap to recompute bandwidth-bound ops, we don't want to end up in a situation
+    # where we have a long chain of pointwise ops from the beginning to the end
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
```

Comment:
A general question about this heuristic, let's say I have a long chain of nodes that I can fuse: n1, n2, ...., n100. Sounds like we should ban the recomputation for n100.

The chain is long, but if we save say n98, then when we recompute n100 in backward pass, we only need recompute n99 and n100. The recomputation cost is not high even though the chain of fusable nodes is long (100 nodes in this case). Then ban recomputation for n100 seems to be not necessary for this specific case? (The example pick an unrealisticaly long chain for discussion purpose)

Reply:
If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

IMO, 2 things:
1. This is kind of a degenerate case, and so I think the precise performance here doesn't matter too much as long as we're not falling into the degenerate case.
2. We do have a heuristic there where we only save it once it narrows down to a "bottleneck".

Reply:
> If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

Yes. Due to saving n98, recomputing n100 is not that expensive even though the fusible chain is very long.